PLANNER-1014 Introduce findbugs and fail build on violations

rsynek opened the pull request:
- findbugs now fails the build
- maxRank set to find violations of "Troubling" severity (findbugs distinguishes Scariest, Scary, Troubling, Of concern)
- current violations have been either fixed or excluded

yurloc commented on optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/descriptor/VariableDescriptor.java:
```diff
@@ -28,7 +29,7 @@
  * @param <Solution_> the solution type, the class with the {@link PlanningSolution} annotation
  */
-public abstract class VariableDescriptor<Solution_> {
+public abstract class VariableDescriptor<Solution_> implements Serializable {
```
Interesting... What is the justification?

rsynek replied:
Inherited classes are used in classes that implement Serializable. Should their fields that are not serializable be rather declared transient?

ge0ffrey replied:
Maybe those casses that use this and implement Serializable shouldn't implement Serializable?

yurloc commented:
> ... integer value between 1 and 20, where 1 to 4 are scariest, 5 to 9 scary, 10 to 14 troubling, and 15 to 20 of concern bugs.

rsynek commented:
@yurloc updated with the rest of changes for other modules. The check should also pass now.

yurloc commented on optaplanner-examples/src/main/java/org/optaplanner/examples/common/persistence/SolutionConverter.java:
```diff
@@ -107,6 +107,10 @@ private SolutionConverter(String dataDirName,
+        if (inputFiles == null) {
+            inputFiles = new File[] {};
+        }
+        
```
Trailing whitespace.

yurloc commented on optaplanner-examples/src/main/java/org/optaplanner/examples/common/persistence/SolutionConverter.java:
```diff
@@ -107,6 +107,10 @@ private SolutionConverter(String dataDirName,
     public void convertAll() {
         File[] inputFiles = inputDir.listFiles();
+        if (inputFiles == null) {
+            inputFiles = new File[] {};
```
Unconventional space after `File[]`.

yurloc commented on optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/result/BenchmarkResultIO.java:
```diff
@@ -92,6 +91,16 @@ public void writePlannerBenchmarkResult(File benchmarkReportDirectory,
+    private File[] retrieveSortedBenchmarkReportDirectories(File benchmarkDirectory) {
+        File[] benchmarkReportDirectories = benchmarkDirectory.listFiles((FileFilter) DirectoryFileFilter.INSTANCE);
+        if (benchmarkReportDirectories == null) {
+            return new File[] {};
```
Unconventional space.

rsynek replied:
Another proof the checkstyle + KIE conventions is necessary. Thanks, will fix

rsynek commented:
@yurloc formatting fixed, feel free to merge.

rsynek commented:
Could we merge this, please?

rsynek commented:
@ge0ffrey could you please review and merge? Thanks.

ge0ffrey commented on optaplanner-core/src/build/findbugs-excludes.xml:
```diff
@@ -0,0 +1,11 @@
+    Uninitialized fields are intentional in this class - to make sure the uninitialized value is the same as JVM
+    would use for uninitialized fields in facts.
+    -->
+    <Class name="org.optaplanner.core.impl.score.director.drools.testgen.fact.TestGenPrimitiveValueProvider"/>
```
Can we do this with annotations instead of an external xml file?

ge0ffrey replied:
it seems there is, but the question is if it introduces an unwelcome dependency (if we make it optional=true, I think it might be ok, if that works for users of course...)
https://stackoverflow.com/a/1829960/472109

yurloc replied:
I think `provided` is what we want - it will be available in compile time but not required for runtime.

ge0ffrey replied:
Either way, I am ok to go with an excludes file for now too.
But let's revert the listFiles() changes.

rsynek replied:
I like the annotation approach, the justification is there right with the code, much more intuitive. The listFiles() won't be reverted - see the other comment.

ge0ffrey commented on optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/move/AbstractMove.java:
```diff
@@ -27,7 +28,7 @@
  * @see Move
  */
-public abstract class AbstractMove<Solution_> implements Move<Solution_> {
+public abstract class AbstractMove<Solution_> implements Move<Solution_>, Serializable {
```
This is dangerous

rsynek replied:
@ge0ffrey I was pretty sure someone will notice this and raise concert during PR review - thanks :) The issue findbugs has is that classes being marked as Serializable should have all the Object fields either Serializable too or transient.

More general question, do we use any classes in OptaPlanner for Java serialization? Or are we forced to declare some classes as Serializable by some framework?

ge0ffrey commented:
jenkins retest this

ge0ffrey commented on optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/result/BenchmarkResultIO.java:
```diff
@@ -78,8 +78,7 @@ public void writePlannerBenchmarkResult(File benchmarkReportDirectory,
         }
-        File[] benchmarkReportDirectories = benchmarkDirectory.listFiles((FileFilter) DirectoryFileFilter.INSTANCE);
-        Arrays.sort(benchmarkReportDirectories);
+        File[] benchmarkReportDirectories = retrieveSortedBenchmarkReportDirectories(benchmarkDirectory);
```
3 lines above there is a "benchmarkDirectory.exists()" check, so this will never return null (unless in a race condition at which time it should fail fast and NPE is ok enough for me).

Can we revert this change and tell Findbugs to exclude it?

yurloc replied:
Oh, isn't this the same situation as the other one in this PR that I already commented on?

If I understand it correctly, the null risk is not with `benchmarkDirectory` but with the result of `listFiles()` which is not guaranteed to be non-null. Replacing it with an empty array looks like an elegant solution that allows the code below to work with it safely without being enclosed with a null-checking `if` statement.

rsynek replied:
@yurloc is correct - it's about the listFiles() result and listFiles() returns null instead of throwing IOException, so we need the null check even if we check for directory. I will leave this null check in place.

ge0ffrey replied:
@rsynek Let's jump on bluejeans to discuss this one

ge0ffrey commented on optaplanner-examples/src/main/java/org/optaplanner/examples/common/persistence/SolutionConverter.java:
```diff
@@ -107,6 +107,10 @@ private SolutionConverter(String dataDirName,
 
     public void convertAll() {
         File[] inputFiles = inputDir.listFiles();
+        if (inputFiles == null) {
```
same principle here: there is already a inputDir.exists() check in the constructor. It's never null, unless there a race condition in which case NPE is fine.

yurloc replied:
This is a null check for the `inputDir.listFiles()` return value.

> Returns null if this abstract pathname does not denote a directory, or if an I/O error occurs.

yurloc replied:
We could perhaps improve the fail-fast approach and also check if `inputDir` is a directory in the constructor.

rsynek replied:
I like the idea of checking if it's really a directory - will change it.

rsynek commented:
@ge0ffrey update with your and @yurloc comments, short summary:
- introduced annotations instead of .xml file for exceptions
- removed the Serializable and marked the field findbugs complained about as `transient`
- the checking of listFiles() result stays, just added a fail fast check for a directory

ge0ffrey commented on optaplanner-core/src/main/java/org/optaplanner/core/impl/partitionedsearch/queue/PartitionChangedEvent.java:
```diff
@@ -29,7 +29,7 @@
     private final long eventIndex;
     private final PartitionChangedEventType type;
-    private final PartitionChangeMove<Solution_> move;
+    private final transient PartitionChangeMove<Solution_> move;
```
Moves aren't Serializable. Sprinkling transient here to make FindBugs happy but not really supporting serializing Moves has no benefit for our users.

Solution proposals
A) Add transient in places to mislead findbugs bug detection. I veto this.
B) Properly make Move's etc Serializable. I think that's unrealistic. That's pandora's box.
C) Remove Serializable from classes that findbugs correctly says aren't really serializable. I like this: no one is serializing moves/descriptors/solvers etc anyway. Yes, this can make CDI-in-the-servlet-context unhappy, but meh.
D) Sprinkle supressWarning annotations or similar Findbugs annotations to avoid findbugs complains. I don't like this much: findbugs should only add to our code in a good way.
E) Exclude the Serializable rule in the findbugs rules to check. I like this one.

Wdyt? I prefer C) or E).

rsynek replied:
For me C) is the best options. Although E) sounds reasonably, excluding the Serializable rule does not make the classes properly Serializable either. I would say we can either do it correctly or should not do it at all.

Is there any particular situation with CDI we need to care about? That would be in my opinion the only reason for E).

rsynek commented:
@ge0ffrey  any more concerns? If not, could you merge it, please? Thanks!

ge0ffrey commented on optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/drools/testgen/fact/TestGenPrimitiveValueProvider.java:
```diff
@@ -15,6 +15,10 @@
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+@SuppressFBWarnings(value = "UR_UNINIT_READ", justification = "Uninitialized fields are intentional in this class - "
+        + "to make sure the uninitialized value is the same as JVM would use for uninitialized fields in facts.")
```
The default values for primitive values are written in the Java Language Spec 1.0 for over than 20 years.
I think it's safe to hard code them :)

ge0ffrey replied:
@yurloc Would you mind if I change this code according to this:


 ```
       if (value.getClass().equals(Byte.class)) {
            uninitialized = (byte) 0;
        } else if (value.getClass().equals(Short.class)) {
            uninitialized = (short) 0;
        } else if (value.getClass().equals(Integer.class)) {
            uninitialized = 0;
        } else if (value.getClass().equals(Long.class)) {
            uninitialized = 0L;
        } else if (value.getClass().equals(Float.class)) {
            uninitialized = 0.0F;
        } else if (value.getClass().equals(Double.class)) {
            uninitialized = 0.0;
        } else if (value.getClass().equals(Boolean.class)) {
            uninitialized = false;
        } else if (value.getClass().equals(Character.class)) {
            uninitialized = (char) 0;
```

yurloc replied:
OK

ge0ffrey commented on pom.xml:
```diff
@@ -101,6 +118,14 @@
+          <configuration>
+            <maxRank>14</maxRank>
+            <includeTests>true</includeTests>
+          </configuration>
```
Does findbugs activate with every build because of this? I don't see that happening?

yurloc replied:
@ge0ffrey No it doesn't. Findbugs plugin execution is part of `fullProfile` defined in `kie-parent` and can be activated by `-Dfull` property. The point of this change is to tweak Findbugs configuration for this project. Without setting maxRank to this specific value, it doesn't any of the problems that this PR fixes.

ge0ffrey commented on pom.xml:
```diff
@@ -80,9 +84,22 @@
+    <dependency>
+      <groupId>com.google.code.findbugs</groupId>
+      <artifactId>annotations</artifactId>
+      <optional>true</optional>
```
I was able to remove this, but getting rid of the need to do those suppresswarning annotations

ge0ffrey commented:
@rsynek I 've pulled everything except for the pom changes into this PR (where I did a number of small changes): https://github.com/kiegroup/optaplanner/pull/384
Please review, so I can merge that.

About the pom changes, let's talk about that on bluejeans. I believe the remaining part should go into kie-parent (unless it's activated by default, which I don't see it doing right now)
